### PR TITLE
Request to add finger-pointing module to ACE 3 mod

### PR DESCRIPTION
**Pull request by TheDrill**
I suggest to add my finger-pointing addon to ACE 3 mod as a module as requested by some players. I tried to adopt the structure of the addon to the standards of ACE 3 as mush as possible. The operability of the resulting module was tested with ACE 3 mod version 1.0.0.3. Please do necessary modifications to the code to make fully compatible with the newest ACE 3 version if needed.

The URL to the thread of original addon:
http://forums.bistudio.com/showthread.php?188895-d_finger-quot-Point-a-finger-at-quot-addon

### CODE REVIEW

**Comment by gienkov** [comment]
:+1:

File: addons/finger/code/fn__init.sqf [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#include "addon.hpp"
+#include "main.hpp"
+
+[] spawn
```
Comment (esteldunedain) [inline addons/finger/code/fn__init.sqf]: spawn is forbidden in ACE

File: addons/finger/code/fn__mainLoop.sqf [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+GVAR(lastFPTime) = -FP_ACTION_TIMEOUT;
+	
+	
+["ACE3 Common", 
+	"point_finger_at", 
+	[LOCALIZE(keyComb), LOCALIZE(keyComb)], 
+	{
+	
+		PV(_rv) = false;
+		if (alive player) then
```
Comment (esteldunedain) [inline addons/finger/code/fn__mainLoop.sqf]: replace player by ACE_player

File: addons/finger/config.cpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
+#include "script_component.hpp"
+
+#define RECOMPILE 0
+
+class CfgPatches
```
Comment (esteldunedain) [inline addons/finger/config.cpp]: macros should be used for CfgPatches entries, like in the rest of the modules

File: addons/finger/config.cpp [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+		};
+		authorUrl = "https://github.com/TheDrill/";
+		
+		addon_version=6;
+		
+		VERSION_CONFIG;
+	};
+};
+
+class CfgFunctions
```
Comment (esteldunedain) [inline addons/finger/config.cpp]: CfgFunctions is not used in ACE in favor of CBA function PREPping

File: addons/finger/stringtable.xml [hunk 5]
```diff
@@ -0,0 +1,34 @@
+﻿<?xml version="1.0" encoding="UTF-8"?>
+<Project name="ace_dfinger">
+  <Package name="strings">
+
+
+
+
+
+
+	<Key ID="STR_Addons_ACE__dfinger__indicator_for_self">
```
Comment (esteldunedain) [inline addons/finger/stringtable.xml]: Key names should use macros to maintaing the same format as on the rest of the modules

File: addons/finger/ACE_Settings.hpp [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#include "code\addon.hpp"
+
+#define LOC(x) $__LOCALIZE_PREFIX__##x
+
+
+class ACE_Settings {
+	class GVAR(FP_indicator_for_self) {
```
Comment (esteldunedain) [inline addons/finger/ACE_Settings.hpp]: The FP prefix should be removed everywhere. Settings should be in camel case like `GVAR(indicatorForSelf)`

File: addons/finger/code/macros.hpp [hunk 7]
```diff
@@ -0,0 +1,54 @@
+#define XOR(x,y) (((x)||(y)) && !((x)&&(y)))
```
Comment (esteldunedain) [inline addons/finger/code/macros.hpp]: Most of this macros are not used an should be removed

File: addons/finger/code/fn__mainLoop.sqf [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+			
+			LOGV(GVAR(pv_fpToServer));
+			
+			if (isServer) then
+			{
+				[0, GVAR(pv_fpToServer)] call GVAR(fnc_pv_fpToServer);
+			}
+			else
+			{
+				publicVariableServer 'GVAR(pv_fpToServer)';
```
Comment (esteldunedain) [inline addons/finger/code/fn__mainLoop.sqf]: This should be replaced by a ACE `globalEvent`

File: addons/finger/code/fn__mainLoop.sqf [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+					random (2*FP_RANDOMIZATION_Y) - FP_RANDOMIZATION_Y,
+					FP_DISTANCE
+				];
+				
+				
+			PV(_dest_players) = [];
+			
+			PV(_pep) = eyePos player;
+			{
+				if (isPlayer _x && _x != player && alive _x &&
```
Comment (esteldunedain) [inline addons/finger/code/fn__mainLoop.sqf]: `isPlayer`should be replaced by `EFUNC(common,isPlayer)`

File: addons/finger/code/fn__comm.sqf [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+					_owner publicVariableClient 'GVAR(pv_fpToClient)';
+				};
+			};			
+		} forEach _dps;
+		
+	};
+		
+	
+	
+	'GVAR(pv_fpToServer)' addPublicVariableEventHandler GVAR(fnc_pv_fpToServer);
```
Comment (esteldunedain) [inline addons/finger/code/fn__comm.sqf]: should be replaced by an ACE event handler

File: addons/finger/code/fn__comm.sqf [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+			
+			if (_owner > 0 && {isPlayer _x}) then
+			{
+				if (player == _x) then
+				{
+					[0, GVAR(pv_fpToClient)] call GVAR(fnc_pv_fpToClient);
+				}
+				else
+				{
+					_owner publicVariableClient 'GVAR(pv_fpToClient)';
```
Comment (esteldunedain) [inline addons/finger/code/fn__comm.sqf]: Replace by ACE event

File: addons/finger/code/fn__comm.sqf [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+{
+	// receives finger-pointing array and broadcasts it to provided players
+	
+	GVAR(fnc_pv_fpToServer) = 
+	{
+		LOGV(_this);
+
+		PV(_val) = _this select 1;
+		
+		PV(_sp) = _val select 0;
```
Comment (esteldunedain) [inline addons/finger/code/fn__comm.sqf]: If possible avoid non-standard macros like this one, which makes the code less readable

File: addons/finger/code/fn__fp_handler.sqf [hunk 13]
```diff
@@ -0,0 +1,44 @@
+#include "addon.hpp"
+#include "main.hpp"
+
+['GVAR(_OEF_renderer)', "onEachFrame", CFUNC(_OEF_renderer)] 
+	call BIS_fnc_addStackedEventHandler;
```
Comment (esteldunedain) [inline addons/finger/code/fn__fp_handler.sqf]: Rendering is better acomplished on a Draw3D event:

```sqf

**Comment by esteldunedain** [comment]
@TheDrill, thanks for submiting this code. I've used The Finger in the past and I really like it. 

To be honest the code still needs a fair amount of work to conform to ACE standards. This is important for maintainability, as having a single code style ensure any dev can quickly understand what's going on if something needs fixing.

In addition to the things I marked above, the following should be fixed:

- Move function code to the `functions` folder

- Minimize the amount of macros defined, and put them in a single place, prefereably `script_components.hpp`

- Function names should conform to standard.

- Function prepping should be done on a CBA preinit EH instead of using BIS function framework.

- Functions are missing proper headings.

- Network communication should be done using the ACE event framework.

- Coding style needs fixing too. Use 4 spaces instead of tabs, open curly braces on the same line.

These things are described on the Wiki section for devs. Feel free to consult here on or the slack chat if in doubt.

Thank you very much

**Comment by TheDrill** [comment]
I see. I was afraid of such "standartization" problems. Unfortunately it is unlikely I would be able to accomplish such massive convertation any time soon. Maybe there would appear a volunteer who would make all necessary changes to the code...

**Comment by PabstMirror** [comment]
I'd be happy to try to convert to CBA/ACE format

**Comment by TheDrill** [comment]
PabstMirror, I wish success to your attempts then!

**Comment by gienkov** [comment]
I think there is already branch with it, named fingering, if you know what
I mean ;).

**Comment by Bummeri** [comment]
http://forums.bistudio.com/showthread.php?188895-d_finger-quot-Point-a-finger-at-quot-addon&p=2938802&viewfull=1#post2938802

I hope you consider adding from the link the LOS check between the finger pointer and the observer so that certain cheap tricks in PvP can be avoided. It has worked very well in our community so far.

**Comment by Glowbal** [comment]
Closed in favor of #1548

**Pull request closed without merging**